// @neo4j/graphql/dist/schema/point.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.cartesianPointDistance = exports.cartesianPointInput = exports.cartesianPoint = exports.pointDistance = exports.pointInput = exports.point = void 0;
const graphql_1 = require("graphql");
exports.point = new graphql_1.GraphQLObjectType({
    name: "Point",
    fields: {
        longitude: {
            type: new graphql_1.GraphQLNonNull(graphql_1.GraphQLFloat),
            resolve: (source) => {
                return source.point.x;
            },
        },
        latitude: {
            type: new graphql_1.GraphQLNonNull(graphql_1.GraphQLFloat),
            resolve: (source) => {
                return source.point.y;
            },
        },
        height: {
            type: graphql_1.GraphQLFloat,
            resolve: (source) => {
                return source.point.z;
            },
        },
        crs: {
            type: new graphql_1.GraphQLNonNull(graphql_1.GraphQLString),
        },
        srid: {
            type: new graphql_1.GraphQLNonNull(graphql_1.GraphQLInt),
            resolve: (source) => {
                return source.point.srid;
            },
        },
    },
});
exports.pointInput = new graphql_1.GraphQLInputObjectType({
    name: "PointInput",
    fields: {
        longitude: {
            type: new graphql_1.GraphQLNonNull(graphql_1.GraphQLFloat),
        },
        latitude: {
            type: new graphql_1.GraphQLNonNull(graphql_1.GraphQLFloat),
        },
        height: {
            type: graphql_1.GraphQLFloat,
        },
    },
});
exports.pointDistance = new graphql_1.GraphQLInputObjectType({
    name: "PointDistance",
    fields: {
        point: {
            type: new graphql_1.GraphQLNonNull(exports.pointInput),
        },
        distance: {
            type: new graphql_1.GraphQLNonNull(graphql_1.GraphQLFloat),
            description: "The distance in metres to be used when comparing two points",
        },
    },
});
exports.cartesianPoint = new graphql_1.GraphQLObjectType({
    name: "CartesianPoint",
    fields: {
        x: {
            type: new graphql_1.GraphQLNonNull(graphql_1.GraphQLFloat),
            resolve: (source) => {
                return source.point.x;
            },
        },
        y: {
            type: new graphql_1.GraphQLNonNull(graphql_1.GraphQLFloat),
            resolve: (source) => {
                return source.point.y;
            },
        },
        z: {
            type: graphql_1.GraphQLFloat,
            resolve: (source) => {
                return source.point.z;
            },
        },
        crs: {
            type: new graphql_1.GraphQLNonNull(graphql_1.GraphQLString),
        },
        srid: {
            type: new graphql_1.GraphQLNonNull(graphql_1.GraphQLInt),
            resolve: (source) => {
                return source.point.srid;
            },
        },
    },
});
exports.cartesianPointInput = new graphql_1.GraphQLInputObjectType({
    name: "CartesianPointInput",
    fields: {
        x: {
            type: new graphql_1.GraphQLNonNull(graphql_1.GraphQLFloat),
        },
        y: {
            type: new graphql_1.GraphQLNonNull(graphql_1.GraphQLFloat),
        },
        z: {
            type: graphql_1.GraphQLFloat,
        },
    },
});
exports.cartesianPointDistance = new graphql_1.GraphQLInputObjectType({
    name: "CartesianPointDistance",
    fields: {
        point: {
            type: new graphql_1.GraphQLNonNull(exports.cartesianPointInput),
        },
        distance: {
            type: new graphql_1.GraphQLNonNull(graphql_1.GraphQLFloat),
        },
    },
});
