// @neo4j/graphql/dist/schema/resolvers/index.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.idResolver = exports.numericalResolver = exports.aggregateResolver = exports.updateResolver = exports.findResolver = exports.deleteResolver = exports.defaultFieldResolver = exports.cypherResolver = exports.createResolver = void 0;
var create_1 = require("./create");
Object.defineProperty(exports, "createResolver", { enumerable: true, get: function () { return __importDefault(create_1).default; } });
var cypher_1 = require("./cypher");
Object.defineProperty(exports, "cypherResolver", { enumerable: true, get: function () { return __importDefault(cypher_1).default; } });
var defaultField_1 = require("./defaultField");
Object.defineProperty(exports, "defaultFieldResolver", { enumerable: true, get: function () { return __importDefault(defaultField_1).default; } });
var delete_1 = require("./delete");
Object.defineProperty(exports, "deleteResolver", { enumerable: true, get: function () { return __importDefault(delete_1).default; } });
var read_1 = require("./read");
Object.defineProperty(exports, "findResolver", { enumerable: true, get: function () { return __importDefault(read_1).default; } });
var update_1 = require("./update");
Object.defineProperty(exports, "updateResolver", { enumerable: true, get: function () { return __importDefault(update_1).default; } });
var aggregate_1 = require("./aggregate");
Object.defineProperty(exports, "aggregateResolver", { enumerable: true, get: function () { return __importDefault(aggregate_1).default; } });
var numerical_1 = require("./numerical");
Object.defineProperty(exports, "numericalResolver", { enumerable: true, get: function () { return __importDefault(numerical_1).default; } });
var id_1 = require("./id");
Object.defineProperty(exports, "idResolver", { enumerable: true, get: function () { return __importDefault(id_1).default; } });
