// @neo4j/graphql/dist/schema/validation/directives.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.writeonlyDirective = exports.uniqueDirective = exports.timestampDirective = exports.relationshipPropertiesDirective = exports.relationshipDirective = exports.readonlyDirective = exports.queryOptions = exports.privateDirective = exports.nodeDirective = exports.idDirective = exports.fulltextDirective = exports.excludeDirective = exports.defaultDirective = exports.cypherDirective = exports.computedDirective = exports.coalesceDirective = exports.aliasDirective = void 0;
const graphql_1 = require("graphql");
const constants_1 = require("../../constants");
const enums_1 = require("./enums");
const scalars_1 = require("./scalars");
exports.aliasDirective = new graphql_1.GraphQLDirective({
    name: "alias",
    description: "Instructs @neo4j/graphql to map a GraphQL field to a Neo4j node or relationship property.",
    locations: [graphql_1.DirectiveLocation.FIELD_DEFINITION],
    args: {
        property: {
            description: "The name of the Neo4j property",
            type: new graphql_1.GraphQLNonNull(graphql_1.GraphQLString),
        },
    },
});
exports.coalesceDirective = new graphql_1.GraphQLDirective({
    name: "coalesce",
    description: "Instructs @neo4j/graphql to wrap the property in a coalesce() function during queries, using the single value specified.",
    locations: [graphql_1.DirectiveLocation.FIELD_DEFINITION],
    args: {
        value: {
            description: "The value to use in the coalesce() function. Must be a scalar type and must match the type of the field with which this directive decorates.",
            type: new graphql_1.GraphQLNonNull(scalars_1.ScalarOrEnumType),
        },
    },
});
exports.computedDirective = new graphql_1.GraphQLDirective({
    name: "computed",
    description: "Informs @neo4j/graphql that a field will be resolved by a custom resolver, and allows specification of any field dependencies.",
    locations: [graphql_1.DirectiveLocation.FIELD_DEFINITION],
    args: {
        from: {
            description: "Fields that the custom resolver will depend on.",
            type: new graphql_1.GraphQLList(new graphql_1.GraphQLNonNull(graphql_1.GraphQLString)),
        },
    },
});
exports.cypherDirective = new graphql_1.GraphQLDirective({
    name: "cypher",
    description: "Instructs @neo4j/graphql to run the specified Cypher statement in order to resolve the value of the field to which the directive is applied.",
    locations: [graphql_1.DirectiveLocation.FIELD_DEFINITION],
    args: {
        statement: {
            description: "The Cypher statement to run which returns a value of the same type composition as the field definition on which the directive is applied.",
            type: new graphql_1.GraphQLNonNull(graphql_1.GraphQLString),
        },
    },
});
exports.defaultDirective = new graphql_1.GraphQLDirective({
    name: "default",
    description: "Instructs @neo4j/graphql to set the specified value as the default value in the CreateInput type for the object type in which this directive is used.",
    locations: [graphql_1.DirectiveLocation.FIELD_DEFINITION],
    args: {
        value: {
            description: "The default value to use. Must be a scalar type and must match the type of the field with which this directive decorates.",
            type: new graphql_1.GraphQLNonNull(scalars_1.ScalarOrEnumType),
        },
    },
});
exports.excludeDirective = new graphql_1.GraphQLDirective({
    name: "exclude",
    description: "Instructs @neo4j/graphql to exclude the specified operations from query and mutation generation. If used without an argument, no queries or mutations will be generated for this type.",
    locations: [graphql_1.DirectiveLocation.INTERFACE, graphql_1.DirectiveLocation.OBJECT],
    args: {
        operations: {
            defaultValue: enums_1.ExcludeOperationEnum.getValues().map((v) => v.value),
            type: new graphql_1.GraphQLNonNull(new graphql_1.GraphQLList(new graphql_1.GraphQLNonNull(enums_1.ExcludeOperationEnum))),
        },
    },
});
exports.fulltextDirective = new graphql_1.GraphQLDirective({
    name: "fulltext",
    description: "Informs @neo4j/graphql that there should be a fulltext index in the database, allows users to search by the index in the generated schema.",
    args: {
        indexes: {
            type: new graphql_1.GraphQLNonNull(new graphql_1.GraphQLList(new graphql_1.GraphQLInputObjectType({
                name: "FullTextInput",
                fields: {
                    name: {
                        type: new graphql_1.GraphQLNonNull(graphql_1.GraphQLString),
                    },
                    fields: {
                        type: new graphql_1.GraphQLNonNull(new graphql_1.GraphQLList(graphql_1.GraphQLString)),
                    },
                    defaultThreshold: {
                        type: graphql_1.GraphQLInt,
                    },
                },
            }))),
        },
    },
    locations: [graphql_1.DirectiveLocation.OBJECT],
});
exports.idDirective = new graphql_1.GraphQLDirective({
    name: "id",
    description: "Indicates that the field is an identifier for the object type. By default; autogenerated, and has a unique node property constraint in the database.",
    locations: [graphql_1.DirectiveLocation.FIELD_DEFINITION],
    args: {
        autogenerate: {
            defaultValue: false,
            type: new graphql_1.GraphQLNonNull(graphql_1.GraphQLBoolean),
        },
        unique: {
            defaultValue: true,
            type: new graphql_1.GraphQLNonNull(graphql_1.GraphQLBoolean),
        },
    },
});
exports.nodeDirective = new graphql_1.GraphQLDirective({
    name: "node",
    description: "Informs @neo4j/graphql of node metadata",
    locations: [graphql_1.DirectiveLocation.OBJECT],
    args: {
        label: {
            description: "Map the GraphQL type to a custom Neo4j node label",
            type: graphql_1.GraphQLString,
        },
        additionalLabels: {
            description: "Map the GraphQL type to match additional Neo4j node labels",
            type: new graphql_1.GraphQLList(new graphql_1.GraphQLNonNull(graphql_1.GraphQLString)),
        },
        plural: {
            description: "Allows for the specification of the plural of the type name.",
            type: graphql_1.GraphQLString,
        },
    },
});
exports.privateDirective = new graphql_1.GraphQLDirective({
    name: "private",
    description: "Instructs @neo4j/graphql to only expose a field through the Neo4j GraphQL OGM.",
    locations: [graphql_1.DirectiveLocation.FIELD_DEFINITION],
});
exports.queryOptions = new graphql_1.GraphQLDirective({
    name: "queryOptions",
    description: "Instructs @neo4j/graphql to inject default values into a query such as a default limit.",
    args: {
        limit: {
            description: "Limit options.",
            type: new graphql_1.GraphQLInputObjectType({
                name: "LimitInput",
                fields: {
                    default: {
                        description: "If no limit argument is supplied on query will fallback to this value.",
                        type: graphql_1.GraphQLInt,
                    },
                    max: {
                        description: "Maximum limit to be used for queries.",
                        type: graphql_1.GraphQLInt,
                    },
                },
            }),
        },
    },
    locations: [graphql_1.DirectiveLocation.OBJECT],
});
exports.readonlyDirective = new graphql_1.GraphQLDirective({
    name: "readonly",
    description: "Instructs @neo4j/graphql to only include a field in generated input type for creating, and in the object type within which the directive is applied.",
    locations: [graphql_1.DirectiveLocation.FIELD_DEFINITION],
});
exports.relationshipDirective = new graphql_1.GraphQLDirective({
    name: "relationship",
    description: "Instructs @neo4j/graphql to treat this field as a relationship. Opens up the ability to create and connect on this field.",
    locations: [graphql_1.DirectiveLocation.FIELD_DEFINITION],
    args: {
        type: {
            type: new graphql_1.GraphQLNonNull(graphql_1.GraphQLString),
        },
        queryDirection: {
            type: enums_1.RelationshipQueryDirectionEnum,
            defaultValue: constants_1.RelationshipQueryDirectionOption.DEFAULT_DIRECTED,
            description: "Valid and default directions for this relationship.",
        },
        direction: {
            type: new graphql_1.GraphQLNonNull(enums_1.RelationshipDirectionEnum),
        },
        properties: {
            type: graphql_1.GraphQLString,
            description: "The name of the interface containing the properties for this relationship.",
        },
    },
});
exports.relationshipPropertiesDirective = new graphql_1.GraphQLDirective({
    name: "relationshipProperties",
    description: "Syntactic sugar to help differentiate between interfaces for relationship properties, and otherwise.",
    locations: [graphql_1.DirectiveLocation.INTERFACE],
});
exports.timestampDirective = new graphql_1.GraphQLDirective({
    name: "timestamp",
    description: "Instructs @neo4j/graphql to generate timestamps on particular events, which will be available as the value of the specified field.",
    locations: [graphql_1.DirectiveLocation.FIELD_DEFINITION],
    args: {
        operations: {
            description: "Which events to generate timestamps on. Defaults to both create and update.",
            defaultValue: enums_1.TimestampOperationEnum.getValues().map((v) => v.value),
            type: new graphql_1.GraphQLNonNull(new graphql_1.GraphQLList(new graphql_1.GraphQLNonNull(enums_1.TimestampOperationEnum))),
        },
    },
});
exports.uniqueDirective = new graphql_1.GraphQLDirective({
    name: "unique",
    description: "Informs @neo4j/graphql that there should be a uniqueness constraint in the database for the decorated field.",
    locations: [graphql_1.DirectiveLocation.FIELD_DEFINITION],
    args: {
        constraintName: {
            description: "The name which should be used for this constraint. By default; type name, followed by an underscore, followed by the field name.",
            type: graphql_1.GraphQLString,
        },
    },
});
exports.writeonlyDirective = new graphql_1.GraphQLDirective({
    name: "writeonly",
    description: "Instructs @neo4j/graphql to only include a field in the generated input types for the object type within which the directive is applied, but exclude it from the object type itself.",
    locations: [graphql_1.DirectiveLocation.FIELD_DEFINITION],
});
