// @neo4j/graphql/dist/schema/validation/enums.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimestampOperationEnum = exports.SortDirection = exports.RelationshipQueryDirectionEnum = exports.RelationshipDirectionEnum = exports.ExcludeOperationEnum = void 0;
const graphql_1 = require("graphql");
const constants_1 = require("../../constants");
exports.ExcludeOperationEnum = new graphql_1.GraphQLEnumType({
    name: "ExcludeOperation",
    values: {
        CREATE: {},
        READ: {},
        UPDATE: {},
        DELETE: {},
    },
});
exports.RelationshipDirectionEnum = new graphql_1.GraphQLEnumType({
    name: "RelationshipDirection",
    values: {
        IN: {},
        OUT: {},
    },
});
exports.RelationshipQueryDirectionEnum = new graphql_1.GraphQLEnumType({
    name: "RelationshipQueryDirection",
    values: {
        [constants_1.RelationshipQueryDirectionOption.DEFAULT_DIRECTED]: {},
        [constants_1.RelationshipQueryDirectionOption.DEFAULT_UNDIRECTED]: {},
        [constants_1.RelationshipQueryDirectionOption.DIRECTED_ONLY]: {},
        [constants_1.RelationshipQueryDirectionOption.UNDIRECTED_ONLY]: {},
    },
});
exports.SortDirection = new graphql_1.GraphQLEnumType({
    name: "SortDirection",
    values: {
        ASC: {},
        DESC: {},
    },
});
exports.TimestampOperationEnum = new graphql_1.GraphQLEnumType({
    name: "TimestampOperation",
    values: {
        CREATE: {},
        UPDATE: {},
    },
});
