// @neo4j/graphql/dist/translate/create-disconnect-and-params.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const create_auth_and_params_1 = __importDefault(require("./create-auth-and-params"));
const constants_1 = require("../constants");
const create_connection_where_and_params_1 = __importDefault(require("./where/create-connection-where-and-params"));
function createDisconnectAndParams({ withVars, value, varName, relationField, parentVar, refNodes, context, labelOverride, parentNode, insideDoWhen, parameterPrefix, }) {
    function createSubqueryContents(relatedNode, disconnect, index) {
        const _varName = `${varName}${index}`;
        const inStr = relationField.direction === "IN" ? "<-" : "-";
        const outStr = relationField.direction === "OUT" ? "->" : "-";
        const relVarName = `${_varName}_rel`;
        const relTypeStr = `[${relVarName}:${relationField.type}]`;
        const subquery = [];
        let params;
        const labels = relatedNode.getLabelString(context);
        const label = labelOverride ? `:${labelOverride}` : labels;
        subquery.push(`WITH ${withVars.join(", ")}`);
        subquery.push(`OPTIONAL MATCH (${parentVar})${inStr}${relTypeStr}${outStr}(${_varName}${label})`);
        const relationship = context.relationships.find((x) => x.properties === relationField.properties);
        const whereStrs = [];
        if (disconnect.where) {
            try {
                const whereAndParams = (0, create_connection_where_and_params_1.default)({
                    nodeVariable: _varName,
                    whereInput: disconnect.where,
                    node: relatedNode,
                    context,
                    relationshipVariable: relVarName,
                    relationship,
                    parameterPrefix: `${parameterPrefix}${relationField.typeMeta.array ? `[${index}]` : ""}.where`,
                });
                if (whereAndParams[0]) {
                    whereStrs.push(whereAndParams[0]);
                }
            }
            catch {
                return { subquery: "", params: {} };
            }
        }
        if (relatedNode.auth) {
            const whereAuth = (0, create_auth_and_params_1.default)({
                operations: "DISCONNECT",
                entity: relatedNode,
                context,
                where: { varName: _varName, node: relatedNode },
            });
            if (whereAuth[0]) {
                whereStrs.push(whereAuth[0]);
                params = { ...params, ...whereAuth[1] };
            }
        }
        if (whereStrs.length) {
            subquery.push(`WHERE ${whereStrs.join(" AND ")}`);
        }
        const preAuth = [parentNode, relatedNode].reduce((result, node, i) => {
            if (!node.auth) {
                return result;
            }
            const [str, p] = (0, create_auth_and_params_1.default)({
                entity: node,
                operations: "DISCONNECT",
                context,
                escapeQuotes: Boolean(insideDoWhen),
                allow: { parentNode: node, varName: _varName, chainStr: `${_varName}${node.name}${i}_allow` },
            });
            if (!str) {
                return result;
            }
            result.disconnects.push(str);
            result.params = { ...result.params, ...p };
            return result;
        }, { disconnects: [], params: {} });
        if (preAuth.disconnects.length) {
            const quote = insideDoWhen ? `\\"` : `"`;
            subquery.push(`WITH ${[...withVars, _varName, relVarName].join(", ")}`);
            subquery.push(`CALL apoc.util.validate(NOT(${preAuth.disconnects.join(" AND ")}), ${quote}${constants_1.AUTH_FORBIDDEN_ERROR}${quote}, [0])`);
            params = { ...params, ...preAuth.params };
        }
        /*
        Replace with subclauses https://neo4j.com/developer/kb/conditional-cypher-execution/
        https://neo4j.slack.com/archives/C02PUHA7C/p1603458561099100
        */
        subquery.push(`FOREACH(_ IN CASE ${_varName} WHEN NULL THEN [] ELSE [1] END | `);
        subquery.push(`DELETE ${_varName}_rel`);
        subquery.push(`)`); // close FOREACH
        // TODO - relationship validation - Blocking, if this were to be enforced it would stop someone from 'reconnecting'
        if (disconnect.disconnect) {
            const disconnects = Array.isArray(disconnect.disconnect) ? disconnect.disconnect : [disconnect.disconnect];
            disconnects.forEach((c, i) => {
                var _a;
                const reduced = Object.entries(c)
                    .filter(([k]) => {
                    var _a;
                    if (k === "_on") {
                        return false;
                    }
                    if (relationField.interface && ((_a = c === null || c === void 0 ? void 0 : c._on) === null || _a === void 0 ? void 0 : _a[relatedNode.name])) {
                        const onArray = Array.isArray(c._on[relatedNode.name])
                            ? c._on[relatedNode.name]
                            : [c._on[relatedNode.name]];
                        if (onArray.some((onKey) => Object.prototype.hasOwnProperty.call(onKey, k))) {
                            return false;
                        }
                    }
                    return true;
                })
                    .reduce((r, [k, v]) => {
                    const relField = relatedNode.relationFields.find((x) => k.startsWith(x.fieldName));
                    const newRefNodes = [];
                    if (relField.union) {
                        Object.keys(v).forEach((modelName) => {
                            newRefNodes.push(context.nodes.find((x) => x.name === modelName));
                        });
                    }
                    else {
                        newRefNodes.push(context.nodes.find((x) => x.name === relField.typeMeta.name));
                    }
                    newRefNodes.forEach((newRefNode) => {
                        const recurse = createDisconnectAndParams({
                            withVars: [...withVars, _varName],
                            value: relField.union ? v[newRefNode.name] : v,
                            varName: `${_varName}_${k}${relField.union ? `_${newRefNode.name}` : ""}`,
                            relationField: relField,
                            parentVar: _varName,
                            context,
                            refNodes: [newRefNode],
                            parentNode: relatedNode,
                            parameterPrefix: `${parameterPrefix}${relField.typeMeta.array ? `[${i}]` : ""}.disconnect.${k}${relField.union ? `.${newRefNode.name}` : ""}`,
                            labelOverride: relField.union ? newRefNode.name : "",
                        });
                        r.disconnects.push(recurse[0]);
                        r.params = { ...r.params, ...recurse[1] };
                    });
                    return r;
                }, { disconnects: [], params: {} });
                subquery.push(reduced.disconnects.join("\n"));
                params = { ...params, ...reduced.params };
                if (relationField.interface && ((_a = c === null || c === void 0 ? void 0 : c._on) === null || _a === void 0 ? void 0 : _a[relatedNode.name])) {
                    const onDisconnects = Array.isArray(c._on[relatedNode.name])
                        ? c._on[relatedNode.name]
                        : [c._on[relatedNode.name]];
                    onDisconnects.forEach((onDisconnect, onDisconnectIndex) => {
                        const onReduced = Object.entries(onDisconnect).reduce((r, [k, v]) => {
                            const relField = relatedNode.relationFields.find((x) => k.startsWith(x.fieldName));
                            const newRefNodes = [];
                            if (relField.union) {
                                Object.keys(v).forEach((modelName) => {
                                    newRefNodes.push(context.nodes.find((x) => x.name === modelName));
                                });
                            }
                            else {
                                newRefNodes.push(context.nodes.find((x) => x.name === relField.typeMeta.name));
                            }
                            newRefNodes.forEach((newRefNode) => {
                                const recurse = createDisconnectAndParams({
                                    withVars: [...withVars, _varName],
                                    value: relField.union ? v[newRefNode.name] : v,
                                    varName: `${_varName}_${k}${relField.union ? `_${newRefNode.name}` : ""}`,
                                    relationField: relField,
                                    parentVar: _varName,
                                    context,
                                    refNodes: [newRefNode],
                                    parentNode: relatedNode,
                                    parameterPrefix: `${parameterPrefix}${relField.typeMeta.array ? `[${i}]` : ""}.disconnect._on.${relatedNode.name}${relField.typeMeta.array ? `[${onDisconnectIndex}]` : ""}.${k}${relField.union ? `.${newRefNode.name}` : ""}`,
                                    labelOverride: relField.union ? newRefNode.name : "",
                                });
                                r.disconnects.push(recurse[0]);
                                r.params = { ...r.params, ...recurse[1] };
                            });
                            return r;
                        }, { disconnects: [], params: {} });
                        subquery.push(onReduced.disconnects.join("\n"));
                        params = { ...params, ...onReduced.params };
                    });
                }
            });
        }
        const postAuth = [parentNode, relatedNode].reduce((result, node, i) => {
            if (!node.auth) {
                return result;
            }
            const [str, p] = (0, create_auth_and_params_1.default)({
                entity: node,
                operations: "DISCONNECT",
                context,
                escapeQuotes: Boolean(insideDoWhen),
                skipRoles: true,
                skipIsAuthenticated: true,
                bind: { parentNode: node, varName: _varName, chainStr: `${_varName}${node.name}${i}_bind` },
            });
            if (!str) {
                return result;
            }
            result.disconnects.push(str);
            result.params = { ...result.params, ...p };
            return result;
        }, { disconnects: [], params: {} });
        if (postAuth.disconnects.length) {
            const quote = insideDoWhen ? `\\"` : `"`;
            subquery.push(`WITH ${[...withVars, _varName].join(", ")}`);
            subquery.push(`CALL apoc.util.validate(NOT(${postAuth.disconnects.join(" AND ")}), ${quote}${constants_1.AUTH_FORBIDDEN_ERROR}${quote}, [0])`);
            params = { ...params, ...postAuth.params };
        }
        subquery.push("RETURN count(*)");
        return { subquery: subquery.join("\n"), params };
    }
    function reducer(res, disconnect, index) {
        if (parentNode.auth) {
            const whereAuth = (0, create_auth_and_params_1.default)({
                operations: "DISCONNECT",
                entity: parentNode,
                context,
                where: { varName: parentVar, node: parentNode },
            });
            if (whereAuth[0]) {
                res.disconnects.push(`WITH ${withVars.join(", ")}`);
                res.disconnects.push(`WHERE ${whereAuth[0]}`);
                res.params = { ...res.params, ...whereAuth[1] };
            }
        }
        res.disconnects.push(`WITH ${withVars.join(", ")}`);
        res.disconnects.push("CALL {");
        if (relationField.interface) {
            const subqueries = [];
            refNodes.forEach((refNode) => {
                const subquery = createSubqueryContents(refNode, disconnect, index);
                if (subquery.subquery) {
                    subqueries.push(subquery.subquery);
                    res.params = { ...res.params, ...subquery.params };
                }
            });
            res.disconnects.push(subqueries.join("\nUNION\n"));
        }
        else {
            const subquery = createSubqueryContents(refNodes[0], disconnect, index);
            res.disconnects.push(subquery.subquery);
            res.params = { ...res.params, ...subquery.params };
        }
        res.disconnects.push("}");
        return res;
    }
    const { disconnects, params } = (relationField.typeMeta.array ? value : [value]).reduce(reducer, {
        disconnects: [],
        params: {},
    });
    return [disconnects.join("\n"), params];
}
exports.default = createDisconnectAndParams;
