// @neo4j/graphql/dist/translate/translate-create.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const create_projection_and_params_1 = __importDefault(require("./create-projection-and-params"));
const create_create_and_params_1 = __importDefault(require("./create-create-and-params"));
const constants_1 = require("../constants");
const create_connection_and_params_1 = __importDefault(require("./connection/create-connection-and-params"));
const create_interface_projection_and_params_1 = __importDefault(require("./create-interface-projection-and-params"));
function translateCreate({ context, node }) {
    var _a, _b, _c, _d, _e, _f;
    const { resolveTree } = context;
    const connectionStrs = [];
    const interfaceStrs = [];
    let connectionParams;
    let interfaceParams;
    const mutationResponse = resolveTree.fieldsByTypeName[node.mutationResponseTypeNames.create];
    const nodeProjection = Object.values(mutationResponse).find((field) => field.name === node.plural);
    const { createStrs, params } = resolveTree.args.input.reduce((res, input, index) => {
        const varName = `this${index}`;
        const create = [`CALL {`];
        const createAndParams = (0, create_create_and_params_1.default)({
            input,
            node,
            context,
            varName,
            withVars: [varName],
            includeRelationshipValidation: true,
        });
        create.push(`${createAndParams[0]}`);
        create.push(`RETURN ${varName}`);
        create.push(`}`);
        res.createStrs.push(create.join("\n"));
        res.params = { ...res.params, ...createAndParams[1] };
        return res;
    }, { createStrs: [], params: {}, withVars: [] });
    let replacedProjectionParams = {};
    let projectionStr;
    let authCalls;
    if (nodeProjection) {
        let projAuth = "";
        const projection = (0, create_projection_and_params_1.default)({
            node,
            context,
            resolveTree: nodeProjection,
            varName: "REPLACE_ME",
        });
        if ((_b = (_a = projection[2]) === null || _a === void 0 ? void 0 : _a.authValidateStrs) === null || _b === void 0 ? void 0 : _b.length) {
            projAuth = `CALL apoc.util.validate(NOT(${projection[2].authValidateStrs.join(" AND ")}), "${constants_1.AUTH_FORBIDDEN_ERROR}", [0])`;
        }
        replacedProjectionParams = Object.entries(projection[1]).reduce((res, [key, value]) => {
            return { ...res, [key.replace("REPLACE_ME", "projection")]: value };
        }, {});
        projectionStr = createStrs
            .map((_, i) => `\nthis${i} ${projection[0]
            // First look to see if projection param is being reassigned
            // e.g. in an apoc.cypher.runFirstColumn function call used in createProjection->connectionField
            .replace(/REPLACE_ME(?=\w+: \$REPLACE_ME)/g, "projection")
            .replace(/\$REPLACE_ME/g, "$projection")
            .replace(/REPLACE_ME/g, `this${i}`)} AS this${i}`)
            .join(", ");
        authCalls = createStrs
            .map((_, i) => projAuth.replace(/\$REPLACE_ME/g, "$projection").replace(/REPLACE_ME/g, `this${i}`))
            .join("\n");
        if ((_d = (_c = projection[2]) === null || _c === void 0 ? void 0 : _c.connectionFields) === null || _d === void 0 ? void 0 : _d.length) {
            projection[2].connectionFields.forEach((connectionResolveTree) => {
                const connectionField = node.connectionFields.find((x) => x.fieldName === connectionResolveTree.name);
                const connection = (0, create_connection_and_params_1.default)({
                    resolveTree: connectionResolveTree,
                    field: connectionField,
                    context,
                    nodeVariable: "REPLACE_ME",
                });
                connectionStrs.push(connection[0]);
                if (!connectionParams)
                    connectionParams = {};
                connectionParams = { ...connectionParams, ...connection[1] };
            });
        }
        if ((_f = (_e = projection[2]) === null || _e === void 0 ? void 0 : _e.interfaceFields) === null || _f === void 0 ? void 0 : _f.length) {
            projection[2].interfaceFields.forEach((interfaceResolveTree) => {
                const relationshipField = node.relationFields.find((x) => x.fieldName === interfaceResolveTree.name);
                const interfaceProjection = (0, create_interface_projection_and_params_1.default)({
                    resolveTree: interfaceResolveTree,
                    field: relationshipField,
                    context,
                    node,
                    nodeVariable: "REPLACE_ME",
                });
                interfaceStrs.push(interfaceProjection.cypher);
                if (!interfaceParams)
                    interfaceParams = {};
                interfaceParams = { ...interfaceParams, ...interfaceProjection.params };
            });
        }
    }
    const replacedConnectionStrs = connectionStrs.length
        ? createStrs.map((_, i) => {
            return connectionStrs
                .map((connectionStr) => {
                return connectionStr.replace(/REPLACE_ME/g, `this${i}`);
            })
                .join("\n");
        })
        : [];
    const replacedInterfaceStrs = interfaceStrs.length
        ? createStrs.map((_, i) => {
            return interfaceStrs
                .map((interfaceStr) => {
                return interfaceStr.replace(/REPLACE_ME/g, `this${i}`);
            })
                .join("\n");
        })
        : [];
    const replacedConnectionParams = connectionParams
        ? createStrs.reduce((res1, _, i) => {
            return {
                ...res1,
                ...Object.entries(connectionParams).reduce((res2, [key, value]) => {
                    return { ...res2, [key.replace("REPLACE_ME", `this${i}`)]: value };
                }, {}),
            };
        }, {})
        : {};
    const replacedInterfaceParams = interfaceParams
        ? createStrs.reduce((res1, _, i) => {
            return {
                ...res1,
                ...Object.entries(interfaceParams).reduce((res2, [key, value]) => {
                    return { ...res2, [key.replace("REPLACE_ME", `this${i}`)]: value };
                }, {}),
            };
        }, {})
        : {};
    const returnStatement = nodeProjection ? `RETURN ${projectionStr}` : "RETURN 'Query cannot conclude with CALL'";
    const cypher = [
        `${createStrs.join("\n")}`,
        authCalls,
        ...replacedConnectionStrs,
        ...replacedInterfaceStrs,
        returnStatement,
    ];
    return [
        cypher.filter(Boolean).join("\n"),
        { ...params, ...replacedProjectionParams, ...replacedConnectionParams, ...replacedInterfaceParams },
    ];
}
exports.default = translateCreate;
